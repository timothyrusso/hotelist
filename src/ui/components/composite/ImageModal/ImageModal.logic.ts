import { useModalsState } from "@/src/ui/state/modals";

export const useImageModal = () => {
	const { modalsActions, modalsSelectors } = useModalsState();

	const closeImageModal = () => {
		modalsActions.hideImageModal();
	};

	return {
		closeImageModal,
		isModalVisible: modalsSelectors.imageModal().visible,
		selectedImage: modalsSelectors.imageModal().image,
	};
};
