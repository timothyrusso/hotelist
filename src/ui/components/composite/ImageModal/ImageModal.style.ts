import { Spacing } from "@/src/ui/constants/style/Spacing";
import { StyleSheet } from "react-native";

export const styles = StyleSheet.create({
	modalOverlay: {
		flex: 1,
		backgroundColor: "rgba(0, 0, 0, 0.8)",
		justifyContent: "center",
		alignItems: "center",
	},
	closeButton: {
		position: "absolute",
		top: Spacing.separator120,
		right: Spacing.Double,
		zIndex: 1,
		width: Spacing.separator40,
		height: Spacing.separator40,
		justifyContent: "center",
		alignItems: "center",
	},
	modalImage: {
		width: "100%",
		height: "100%",
		borderRadius: Spacing.Fourfold,
	},
});
