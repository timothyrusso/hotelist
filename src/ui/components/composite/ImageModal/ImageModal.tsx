import { Platforms } from "@/src/ui/constants/platform/Platforms";
import { Colors } from "@/src/ui/constants/style/Colors";
import { Spacing } from "@/src/ui/constants/style/Spacing";
import { Image } from "expo-image";
import { Modal, View } from "react-native";
import { CloseButton } from "../../basic/CloseButton/CloseButton";
import { useImageModal } from "./ImageModal.logic";
import { styles } from "./ImageModal.style";

export const ImageModal = () => {
	const { isModalVisible, closeImageModal, selectedImage } = useImageModal();

	return (
		<Modal
			visible={isModalVisible}
			transparent
			animationType="fade"
			onRequestClose={closeImageModal}
			statusBarTranslucent={Platforms.Android}
		>
			<View style={styles.modalOverlay}>
				<CloseButton
					onPress={closeImageModal}
					color={Colors.white.default}
					style={styles.closeButton}
					size={Spacing.Fourfold}
				/>
				{selectedImage ? (
					<Image
						source={{ uri: selectedImage }}
						style={styles.modalImage}
						contentFit="contain"
					/>
				) : null}
			</View>
		</Modal>
	);
};
