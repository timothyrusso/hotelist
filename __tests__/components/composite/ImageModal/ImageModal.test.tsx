import { ImageModal } from "@/src/ui/components/composite/ImageModal/ImageModal";
import { fireEvent, render } from "@testing-library/react-native";
import type { StyleProp, ViewStyle } from "react-native";

// Mock child components and dependencies
jest.mock("@/src/ui/components/basic/CloseButton/CloseButton", () => {
	// eslint-disable-next-line @typescript-eslint/no-require-imports
	const { Pressable, Text } = require("react-native");
	return {
		CloseButton: ({
			onPress,
			color,
			style,
			size,
		}: {
			onPress: () => void;
			color?: string;
			style?: StyleProp<ViewStyle>;
			size?: number;
		}) => (
			<Pressable testID="close-button" onPress={onPress} style={style}>
				<Text testID="close-button-text">
					Close ({color}, {size})
				</Text>
			</Pressable>
		),
	};
});

// Mock the custom hook to control modal state and behavior
jest.mock("@/src/ui/components/composite/ImageModal/ImageModal.logic", () => ({
	useImageModal: jest.fn(),
}));

// Mock expo-image to avoid native image loading and test image props
jest.mock("expo-image", () => ({
	Image: ({
		source,
		style,
		contentFit,
	}: {
		source: { uri: string };
		style: StyleProp<ViewStyle>;
		contentFit: string;
	}) => {
		// eslint-disable-next-line @typescript-eslint/no-require-imports
		const { View, Text } = require("react-native");
		return (
			<View testID="modal-image" style={style}>
				<Text testID="image-uri">{source.uri}</Text>
				<Text testID="image-content-fit">{contentFit}</Text>
			</View>
		);
	},
}));

describe("ImageModal", () => {
	// Get reference to the mocked hook for controlling test scenarios
	const mockUseImageModal =
		// eslint-disable-next-line @typescript-eslint/no-require-imports
		require("@/src/ui/components/composite/ImageModal/ImageModal.logic").useImageModal;

	beforeEach(() => {
		jest.clearAllMocks();
	});

	describe("snapshot tests", () => {
		it("should render correctly when modal is visible with image", () => {
			mockUseImageModal.mockReturnValue({
				isModalVisible: true,
				closeImageModal: jest.fn(),
				selectedImage: "https://example.com/test-image.jpg",
			});

			const { toJSON } = render(<ImageModal />);
			expect(toJSON()).toMatchSnapshot();
		});

		describe("functional tests", () => {
			it("should render modal when isModalVisible is true", () => {
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: "https://example.com/test-image.jpg",
				});

				const { getByTestId } = render(<ImageModal />);
				expect(getByTestId("close-button")).toBeTruthy();
			});

			it("should not render modal content when isModalVisible is false", () => {
				mockUseImageModal.mockReturnValue({
					isModalVisible: false,
					closeImageModal: jest.fn(),
					selectedImage: "https://example.com/test-image.jpg",
				});

				const { queryByTestId } = render(<ImageModal />);
				expect(queryByTestId("close-button")).toBeNull();
			});

			it("should render image when selectedImage is provided", () => {
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: "https://example.com/test-image.jpg",
				});

				const { getByTestId } = render(<ImageModal />);
				expect(getByTestId("modal-image")).toBeTruthy();
				expect(getByTestId("image-uri").children[0]).toBe(
					"https://example.com/test-image.jpg",
				);
			});

			it("should not render image when selectedImage is null", () => {
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: null,
				});

				const { queryByTestId } = render(<ImageModal />);
				expect(queryByTestId("modal-image")).toBeNull();
			});

			it("should call closeImageModal when close button is pressed", () => {
				// Create a mock function to track if closeImageModal is called
				const mockCloseImageModal = jest.fn();
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: mockCloseImageModal,
					selectedImage: "https://example.com/test-image.jpg",
				});

				const { getByTestId } = render(<ImageModal />);
				fireEvent.press(getByTestId("close-button"));

				expect(mockCloseImageModal).toHaveBeenCalledTimes(1);
			});

			it("should render CloseButton with correct props", () => {
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: "https://example.com/test-image.jpg",
				});

				const { getByTestId } = render(<ImageModal />);
				const closeButtonText = getByTestId("close-button-text");
				expect(closeButtonText.children[0]).toContain("Close");
			});

			it("should render image with correct contentFit", () => {
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: "https://example.com/test-image.jpg",
				});

				const { getByTestId } = render(<ImageModal />);
				expect(getByTestId("image-content-fit").children[0]).toBe("contain");
			});
		});

		describe("edge cases", () => {
			it("should handle very long image URL", () => {
				// Test with extremely long URL to ensure no rendering issues
				const longUrl = "https://example.com/" + "a".repeat(1000) + ".jpg";
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: longUrl,
				});

				const { getByTestId } = render(<ImageModal />);
				expect(getByTestId("modal-image")).toBeTruthy();
				expect(getByTestId("image-uri").children[0]).toBe(longUrl);
			});

			it("should handle undefined selectedImage", () => {
				// Test behavior when selectedImage is undefined (should not render image)
				mockUseImageModal.mockReturnValue({
					isModalVisible: true,
					closeImageModal: jest.fn(),
					selectedImage: undefined,
				});

				const { queryByTestId } = render(<ImageModal />);
				expect(queryByTestId("modal-image")).toBeNull();
			});
		});
	});
});
