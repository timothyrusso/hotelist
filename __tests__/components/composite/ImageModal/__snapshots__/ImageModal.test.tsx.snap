// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageModal snapshot tests should render correctly when modal is visible with image 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  statusBarTranslucent={false}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.8)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "position": "absolute",
          "right": 12,
          "top": 120,
          "width": 40,
          "zIndex": 1,
        }
      }
      testID="close-button"
    >
      <Text
        testID="close-button-text"
      >
        Close (
        #FFFFFF
        , 
        24
        )
      </Text>
    </View>
    <View
      style={
        {
          "borderRadius": 24,
          "height": "100%",
          "width": "100%",
        }
      }
      testID="modal-image"
    >
      <Text
        testID="image-uri"
      >
        https://example.com/test-image.jpg
      </Text>
      <Text
        testID="image-content-fit"
      >
        contain
      </Text>
    </View>
  </View>
</Modal>
`;
